src/app/drive/components/ShareInviteDialog/ShareInviteDialog.tsx

```diff
@@ -82,185 +97,214 @@ const ShareInviteDialog = (props: ShareInviteDialogProps): JSX.Element => {
   };
 
-  //TODO: EXTRACT THIS LOGIC OUT OF THE DIALOG
-  const onInvite = async () => {
-    const sharingPromises = [] as AsyncThunkAction<string | void, ShareFileWithUserPayload, { state: RootState }>[];
+  const getUserPublicKey = async (email: string): Promise<string> => {
+    let publicKey = '';
+    try {
+      const publicKeyResponse = await userService.getPublicKeyByEmail(email);
+      publicKey = publicKeyResponse.publicKey;
+    } catch (error) {
+      if (error instanceof Error) {
+        const errorBody = JSON.parse(error.message);
+        if (errorBody.statusCode !== 404) {
+          throw error;
+        }
+      }
+    }
+    return publicKey;
+  };
 
-    let trackingRestrictedSharedProperties: TrackingPlan.RestrictedSharedProperties;
+  const processInvites = async (usersToInvite: UsersToInvite[]) => {
+    const sharingPromises = [] as AsyncThunkAction<string | void, ShareFileWithUserPayload, { state: RootState }>[];
 
-    if (usersToInvite.length === 0 && isValidEmail(email)) {
-      const userRoleId = props.roles.find((role) => role.name === userRole)?.id;
+    usersToInvite.forEach((user) => {
+      const userRoleId = props.roles.find((role) => role.name === user.userRole)?.id;
       if (!userRoleId) return;
 
       sharingPromises.push(
         dispatch(
           sharedThunks.shareItemWithUser({
-            encryptionAlgorithm: props.itemToShare.encryptVersion,
+            encryptionAlgorithm: props.itemToShare.encryptionAlgorithm,
             itemId: props.itemToShare.uuid,
             itemType: props.itemToShare.isFolder ? 'folder' : 'file',
             roleId: userRoleId,
-            sharedWith: email,
+            sharedWith: user.email,
             notifyUser,
             notificationMessage: messageText,
+            publicKey: user.publicKey,
+            isNewUser: user.isNewUser,
           }),
         ),
       );
+    });
+    const trackingRestrictedSharedProperties: TrackingPlan.RestrictedSharedProperties = {
+      is_folder: props.itemToShare?.isFolder,
+      share_type: 'private',
+      user_id: props.itemToShare?.userId,
+      item_id: props.itemToShare?.id,
+      invitations_send: 1,
+    };
 
-      trackingRestrictedSharedProperties = {
-        is_folder: props.itemToShare.item.isFolder,
-        share_type: 'private',
-        user_id: props.itemToShare.item.userId,
-        item_id: props.itemToShare.item.id,
-        invitations_send: 1,
-      };
-    } else {
-      usersToInvite.forEach((user) => {
-        const userRoleId = props.roles.find((role) => role.name === user.userRole)?.id;
-        if (!userRoleId) return;
+    await Promise.all(sharingPromises);
+    trackRestrictedShared(trackingRestrictedSharedProperties);
+  };
+
+  //TODO: EXTRACT THIS LOGIC OUT OF THE DIALOG
+  const onInvite = async ({ inviteNewUsers = false }) => {
+    const preCreateUsers = inviteNewUsers;
+    const usersList = [...usersToInvite];
+    let isThereAnyNewUser = newUsersExists;
```

Comment:
Same here, maybe we should rename the name of the state.

Reply:
No, this `let` is going to change in this function flow and it is essencial to cover the two possible invitation cases we have, I prefer to not set any state here  to prevent any rerendering while this is being executed (Maybe now that we are using React 18 the possibility of rendering problems is nonexistent tho)

```
   if (usersList.length === 0 && isValidEmail(email)) {
      const publicKey = await getUserPublicKey(email);
      if (!publicKey && !preCreateUsers) {
        isThereAnyNewUser = true;
      }
      usersList.push({
        email,
        userRole,
        isNewUser: !publicKey,
        publicKey,
      });
    }
```